apis/r/R/SOMASparseNDArray.R

```diff
@@ -63,9 +63,13 @@ SOMASparseNDArray <- R6::R6Class(
     #' [`TsparseMatrix`][`Matrix::TsparseMatrix-class`]. Character dimension
     #' names are ignored because `SOMANDArray`'s use integer indexing.
     #'
-    write = function(values) {
+    write = function(values, bbox = NULL) {
       stopifnot(
-        "'values' must be a matrix" = is_matrix(values)
+        "'values' must be a matrix" = is_matrix(values),
+        is.null(bbox) || length(bbox) == length(dim(values)),
+        is.null(bbox) ||
+          (is_integerish(bbox) || bit64::is.integer64(bbox)) ||
+          (is.list(bbox) && vapply_lgl(bbox, function(x, n) length(x) == n, n = length(dim(values))))
```

Comment:
split this up into multiple `stopifnot`. the "must be a matrix" is just one error message covering a multitude of possible errors.

i think we need three stopifnots:

* is a matrix
* null or lengths match
* integer types; dimensions

Reply:
This is not the case. The `'values' must be a matrix` message occurs only if `is_matrix(values)` returns `FALSE`; the other conditions don't (yet) have an informative message
```
>  uri <- withr::local_tempdir("sparse-ndarray-bbox")
>  nrows <- 100L
>  ncols <- 500L
>  ndarray <- SOMASparseNDArrayCreate(uri, type = arrow::int32(), shape = c(nrows, ncols))
>  mat <- create_sparse_matrix_with_int_dims(nrows, ncols)
>  ndarray$write(mat, bbox = TRUE) # violation of the second test
Error in ndarray$write(mat, TRUE) :
is.null(bbox) || length(bbox) == length(dim(values)) is not TRUE
```
Informative error messages are needed, and I will add them, but it is already "multiple `stopifnots`"

Reply:
Informative error messages added in 9b3ea56